### PR TITLE
Make test_make_sparse_regression more stable and explain the non-determinism of SparseRegressionDataGen

### PR DESCRIPTION
**Pull request by lijinf2**
Issue: https://github.com/NVIDIA/spark-rapids-ml/issues/892
Relevant PR: https://github.com/NVIDIA/spark-rapids-ml/pull/742

### CODE REVIEW

**Comment by lijinf2** [comment]
build

File: python/benchmark/gen_data_distributed.py [hunk 1]
```diff
@@ -580,6 +580,11 @@ def make_regression_udf(iter: Iterable[pd.DataFrame]) -> Iterable[pd.DataFrame]:
 class SparseRegressionDataGen(DataGenBaseMeta):
     """Generate sparse regression dataset using a distributed version of sklearn.datasets.regression,
     including features and labels.
+
+    SparseRegressionDataGen is a non-deterministic data generator that may produce different DataFrames across multiple runs.
+    While setting a random state ensures that the generated DataFrame will vary between runs, it does not guarantee reproducibility of the same output for the same random state.
+    This is because each partition utilizes a combination of the random state and the partition index as its local seed.
+    However, partition indices are not guaranteed to be continuous or deterministic, leading to variability even when the random state is provided.
```
Comment (eordentlich) [inline python/benchmark/gen_data_distributed.py:587]: We have seen they can be non-contiguous.   Why not deterministic?

Comment (lijinf2) [reply python/benchmark/gen_data_distributed.py:587]: Printed out generated y non-deterministic. 
Need further investigation.

**Comment by lijinf2** [comment]
build

File: python/benchmark/gen_data_distributed.py (hunk 1 above)
Comment (eordentlich) [reply python/benchmark/gen_data_distributed.py]: Good catch.   So is it still non-deterministic?

File: python/benchmark/gen_data_distributed.py [hunk 2]
```diff
@@ -580,6 +592,12 @@ def make_regression_udf(iter: Iterable[pd.DataFrame]) -> Iterable[pd.DataFrame]:
 class SparseRegressionDataGen(DataGenBaseMeta):
     """Generate sparse regression dataset using a distributed version of sklearn.datasets.regression,
     including features and labels.
+
+    SparseRegressionDataGen is a deterministic data generator when running in Spark local mode. It produces
+    identical DataFrames when initialized with the same random state. However, in Spark standalone mode, it may occasionally
+    yield different DataFrames even with the same random state. This is because each partition derives its seed from
+    a combination of the provided random state and the partition index. Since partition indices are not guaranteed to be continuous
+    or consistent across runs in standalone mode, this can lead to non-deterministic output despite a fixed random state.
```
Comment (eordentlich) [inline python/benchmark/gen_data_distributed.py:600]: Should be consistent as long as number of partitions is the same but that could change depending on number of executor-cores for example.    Maybe need to give option to set the num partitions argument to spark.range , the base dataframe/rdd.

**Comment by lijinf2** [comment]
build

File: python/benchmark/gen_data_distributed.py (hunk 1 above)
Comment (lijinf2) [reply python/benchmark/gen_data_distributed.py]: Revised the comment.

File: python/benchmark/gen_data_distributed.py (hunk 2 above)
Comment (lijinf2) [reply python/benchmark/gen_data_distributed.py]: Added a warning message when output_num_files not specified.

File: python/benchmark/gen_data_distributed.py [hunk 3]
```diff
@@ -785,7 +802,6 @@ def make_sparse_regression_udf(
                     random_state=generator,
                     format="csr",
                     dtype=dtype,
-                    data_rvs=np.random.randn,
```
Comment (eordentlich) [inline python/benchmark/gen_data_distributed.py:788]: Curious why this was deleted?

File: python/benchmark/gen_data_distributed.py [hunk 4]
```diff
@@ -816,10 +831,7 @@ def make_sparse_regression_udf(
                     sparse_matrix = sparse_matrix[:, informative_shuffle_indices]
                 informative = sparse_matrix[:, :n_informative]
 
-                if use_cupy:
-                    redundant_mul = np.random.rand(n_informative, redundant_cols)
```
Comment (eordentlich) [inline python/benchmark/gen_data_distributed.py:820]: Was this a bug to use np vs cp here?

Comment (lijinf2) [reply python/benchmark/gen_data_distributed.py:820]: Not a bug. redundant_mul must be a numpy array because it is used to construct a CPU-based SciPy sparse matrix, regardless of whether use_cupy is true or false. cupy is only utilized for generating the label column.

The cupyx sparse matrix is not used here, which may be fine for now.

File: python/benchmark/gen_data_distributed.py (hunk 3 above)
Comment (lijinf2) [reply python/benchmark/gen_data_distributed.py:788]: For the function to reuse generator which was initialized with random_state. 
Keeping it here will instruct the function to use np.random.randn, which creates non-determinism and it was not initialized with random_state.

Comment (eordentlich) [reply python/benchmark/gen_data_distributed.py:788]: Does the distribution change from normal to uniform 0,1?  To keep distribution could an initialized generator function be passed in here?

Comment (lijinf2) [reply python/benchmark/gen_data_distributed.py:788]: Good catch. 

Just added back the data_rvs per np.random had been revised to be initialized with the partition seed.

**Comment by lijinf2** [comment]
build

File: python/benchmark/gen_data_distributed.py [hunk 5]
```diff
@@ -807,7 +824,6 @@ def make_sparse_regression_udf(
                         random_state=generator,
                         format="csr",
                         dtype=dtype,
-                        data_rvs=np.random.randn,
```
Comment (eordentlich) [inline python/benchmark/gen_data_distributed.py:810]: Here too looks like.

**Comment by lijinf2** [comment]
build

**Review by eordentlich — APPROVED** [review]
👍
